Immutable :value options

kellyredding opened the pull request:
Specifying an option with the `:value` rule makes it immutable.

``` ruby
option :whatever, :value => 'idontcare'

thing.whatever           #=> 'idontcare'
thing.whatever = 'huh'   #=> NoMethodError ...
thing.whatever 'huh'     #=> ArgumentError ...
```

Addresses #57.

kellyredding committed:
design writeup in README for immutable "value" options

kellyredding committed:
more efficient implementation of Namespace#apply

This one makes sure to use the (dynamic) writers.  Eat your own
dogfood please.

kellyredding committed:
reorganized the rules unit tests on the Option class

kellyredding committed:
make an Option immutable if created with a `:value` rule

kellyredding committed:
Have a namespace not respond to and appropriately handle value options

Needs to raise NoMethodError if trying to write to a :value option.
Needs to raise ArgumentError if trying to write a :value option using
the reader w/ some args.

kellyredding commented:
@jcredding - take a look at this.  I still want to add a few system tests to validate this trumps the `:default` rule, works with lazy procs, etc.  But its good to start reviewing.

kellyredding commented:
(note to self)

Need to verify this still allows apply to work (I don't think it does).  May need to not have f7e4a84 after all.

kellyredding commented:
(note to self)

Maybe there is a way to configure a proxy/ns to be all values.  Doing this would insert a `:value` rule when adding options (whether pre-defined or dynamic).  Proxy would configure this on any recursive namespaces.

kellyredding committed:
some refactorings to the namespace class

jcredding commented on lib/ns-options/namespace.rb:
```diff
@@ -26,32 +26,48 @@ def namespace(name, &block)
+      super
     end
 
     def method_missing(meth, *args, &block)
```
@kellyredding - I believe you added an issue for this, and I agree, it'd be nice if we could do something here.

jcredding commented on lib/ns-options/namespace.rb:
```diff
@@ -59,10 +75,23 @@ def required_set?;        @__data__.required_set?;        end
 
+    # The opposite of #to_hash. Takes a hash representation of options and
+    # namespaces and mass assigns option values.
+    def apply(values=nil)
```
@kellyredding - Why did this move from the data class to here? Guess it doesn't really matter, just curious.

kellyredding replied:
@jcredding No good reason.  It should be on the data class since that is where the data lives.  I'll move it back.  Good catch.

jcredding commented:
@kellyredding - Overall looking good, think you're on the write track.

kellyredding commented:
Note:

A big change here is the `apply` method on a namespace/proxy now uses the same logic as an option writer.  So if you try to apply data to a value option, you'll get a `NoMethodError`.

kellyredding committed:
added another value option usage example

Illustrates trying to write an option value using a DSL-style call.
It will raise an ArgumentError.

kellyredding committed:
refactored the Opiton, Options, and Namespaces classes.

Options and Namespaces classes now proxy a hash instead of
subclassing one.  Better pattern, more extendable.

The Option class expects the 3 params it needs already parsed.  This
will allows for things to inject rules and other args when building
new options.  Had to go rework the setup of most of the Option
tests to make them send 3 args always.

The Options class now parses the args directly.  The `args`
singleton method parses the args and returns them as an array
in the order the Option constructor expects them in (reverse of
previous order).  Had to rework Options and AssertMacros to
handle the args in this new reversed order.  Going to inject a
:value rule here when mass handling :value options.

Also, added the concept of an `option_type` to the options class.
This will allow for specialized handling of options for different
types.  Going to be used in mass handling :value options.

kellyredding committed:
have a custom error be thrown when writing to value options

kellyredding committed:
pending value handling for the Option and Options classes

Pending values allow for options to be set as immutable in mass and
have those immutable values be applied on the fly.  If an option is
pending a value, the first thing written to it will be its immutable
value.

kellyredding committed:
:values handling for namespaces

Namespaces can be created with the `:values` flag and this will
make them create all child options as immutable :value options. All
child namespaces will do the same.

kellyredding committed:
accessing immutables raises custom exception
